cdc/sorter/leveldb/table_sorter.go

```
@@ -0,0 +1,539 @@
+		// output resolved events as much as possible.
+		ls.outputBufferedResolvedEvents(state.outputBuf, true)
+		lenResolvedEvents, _ = state.outputBuf.len()
+	}
+	// New received events.
+	newEvents := state.eventsBuf[:n]
+
+	// It can only acquire a snapshot when
+	// 1. No buffered resolved events, they must be sent before
+	//    sending further resolved events from snapshot.
+	needSnap := lenResolvedEvents == 0
+	// 2. There are some events that can be resolved.
+	//    -------|-----------------|-------------|-------> time
+	//    exhaustedResolvedTs
+	//                        maxCommitTs
+	//                                    maxResolvedTs
+	//    -------|-----------------|-------------|-------> time
+	//    exhaustedResolvedTs
+	//                       maxResolvedTs
+	//                                      maxCommitTs
+	needSnap = needSnap && state.exhaustedResolvedTs < maxCommitTs
+
+	// Write new events and delete sent keys.
+	snapCh, err :=
+		ls.asyncWrite(ctx, newEvents, state.outputBuf.deleteKeys, needSnap)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Reset buffer as delete keys are scheduled.
+	state.outputBuf.resetDeleteKey()
+	// Try shrink buffer to release memory.
+	state.outputBuf.maybeShrink()
+
+	if !needSnap {
+		// No new events and no buffered resolved events.
+		// -------|-----------------|-------------|-------> time
+		//   maxCommitTs
+		//                 exhaustedResolvedTs
+		//                                   maxResolvedTs
+		if maxCommitTs <= state.exhaustedResolvedTs && maxResolvedTs != 0 &&
+			lenResolvedEvents == 0 {
+			ls.outputResolvedTs(maxResolvedTs)
+		}
+		return nil
+	}
+
+	// Wait for writing and deleting.
+	var snap message.LimitedSnapshot
+	var ok bool
+	select {
+	case <-ctx.Done():
+		return errors.Trace(ctx.Err())
+	case snap, ok = <-snapCh:
+	}
+	if !ok {
+		if needSnap {
+			log.Panic("need snapshot", zap.Uint64("tableID", ls.tableID))
+		}
+	}
```

Could this happen or not?

---

No, if there is no bug. It's a kind of assert for preventing bugs.

---

Why not return an error and terminate the changefeed permanently, panic won't pause the changefeed, only record a fatal log and node restart.